### ISSUE TITLE
API stabilization

### ISSUE THREAD

**Issue by ethomson** [libgit2/discussions#4]
@tschoonj has (quite rightly) [complained that our API is not very stable](https://github.com/libgit2/libgit2.github.com/pull/82).  We've used the excuse that we're "not 1.0" and that once we were, then we would guarantee some stability in the API.  We've also used the excuse that most of our consumers are bindings (and most of them are maintained by us) that this isn't really such a big problem.

But given that we've been talking about this for several _years_ now, it feel like a pretty weak answer.

As a maintainer, it's easy to lose sight of these problems.  I'm glad that @tschoonj reminded us that this is a problem, because quite honestly, we get _very little feedback_ about these changes.  This means that there are quite likely to be people sitting around cursing our names under their breath.

At the [summit](https://github.com/libgit2/discussions/issues/3), we discussed briefly some strategies for moving forward here.  Let's pick that thread back up - obviously, we need a plan and we need to start being more responsible.

**Comment by ilovezfs** [comment]
One specific pain point here is [libgit2-glib](https://github.com/GNOME/libgit2-glib), which seems to break each time there is a libgit2 upgrade, and fixes are rarely in libgit2-glib master yet, let alone in a stable released version.

**Comment by tiennou** [comment]
Maybe integrating something like https://github.com/lvc/abi-compliance-checker into CI would help in that regard ? Well, would you look at [that](https://abi-laboratory.pro/?view=timeline&l=libgit2) ✨.

**Comment by tiennou** [comment]
Yeah, integrating into CI might be a little heavy-handed, that was just a proposal for tracking those ABI breaks. Maybe a bunch of scripts that perform those checks when preparing a release would be sufficient at catching regressions…

We should also start a list of those brittle APIs we are aware of, so we're not caught with too many of them (or v2 might happen sooner 😜). Right now, I know of :

- the remote creation API (cleanup in https://github.com/libgit2/libgit2/pull/4667). This is going to make it harder to tweak remote creation.

- the ENV-conforming changes of https://github.com/libgit2/libgit2/pull/4856. This is going to make it impossible to add support for additional envvars.

- structifying some of the "internal" fetch machinery (uncovered in https://github.com/libgit2/libgit2/pull/4747, `fetch: pass negotiation data around as a struct`). I don't expect the shallow support to be ready for 1.0, but maybe cherry-picking just that commit so we don't freeze an inextensible API would be *hem* saner 😉.

- the commit creation functions of https://github.com/libgit2/libgit2/pull/3075.

To be clear, most of those PRs are both API-proofing *and* feature-additions. Extracting the API-proofing parts might make them more amenable to be ready for 1.0 ? If you're aware of some other parts, I might be able to concentrate getting those ready.

Also, for clarification purposes, I ended up confusing myself with 1.0 being ABI-compatible with 0.27, which doesn't make sense wrt. semver. Is the plan to be as compatible as possible with 0.27, while still keeping the amount of breakage sensible/having a deprecation path available ?

**Comment by tiennou** [comment]
(For consistency, I'm moving the discussion over from libgit2/libgit2#4316 to here, because IMHO it makes more sense to discuss the general aspects of it here than in a somewhat unrelated PR, and we're touching on the very interesting concept of how to actually handle changes in our various options structures.)

So, I don't think the story is complete from the perspective of users. Let's say for the sake of argument that all internal changes are made "correctly" (to be defined later 😉) w.r.t to the irrelevant fix. It's also likely that we'll have to have the "struct of `int version` + unions" internally anyway so we don't run afoul of the Laws Of The Aliasing.

Now, what does the users' code sees ? Looking at ABI compat, "old" linked code will provide a "deprecated" version number, which will be auto-upgraded internally (using the struct-of-unions), so there should be no crash or weirdness (which is good, because user code don't have to recompile).

On the other hand, the API compatibility is much less clear :

1) If the user recompiles, it will now see the new struct layout (because it's likely just initializing its `git_system_options` struct via a `GIT_SYSTEM_OPTIONS_INIT`), and if it was depending upon a changed option will likely now see an error.

2) Handling case 1 means that the `git_system_options` logical layout must be preserved. If we make the struct-of-unions public, all members will need to be updated in the users' code (ie. `opts.force` becomes `opts.v1.force`), thus, an error.

3) In the public API, we might now provide a `git_system_options` (the API-compatible version), as well as `git_system_options_v2` (the version with a fix). Which one do we use in the function declarations ? If we go the API-compatibility way, the declaration must be kept as `git_system_options`, which means that "new" users that want the fix have to live with "invalid type to a parameter of type `git_system_options`"-style warnings. We also now have 3 different views of the options structure : the internal one, and the 2 public versions.

4) We could also do the reverse, and consider that the `git_system_options` struct to always mean the latest version, and add the `_v1` tag to the old one. Which means that users that recompile against a semver-compatible version now need to change their code to the "legacy version" (ie. `git_system_options_v1`), and live with the warnings.

I think this is a nice round up of all cases, but feel free to describe ones I missed if you feel like they're not handled above.

So, my ~final~current understanding of all of this is that, unless we choose to do case 3 or 4, *we can never modify an option structures' members, unless it's accompanied by a MAJOR version bump*. This might seem obvious, but at least to me it wasn't that clear-cut, and I feel like it makes any kind of attempt at handling BC-break futile.
